### PR TITLE
[REVIEW] Fix issues with building cudf in a non-conda environment

### PR DESCRIPTION
**Pull request by galipremsagar**
Extension to https://github.com/rapidsai/rmm/pull/733

Also resolves: https://github.com/rapidsai/cudf/issues/7578

This PR introduces fixes to `setup.py` in cudf build and also adds a DockerFile with which one can create a container with cudf fully built from source. Arrow & rmm will also be built from the source. This PR also contains a `requirements.txt` which has packages needed by cudf and that are available to install via pip.

### CODE REVIEW

**Comment by galipremsagar** [comment]
Validating install of last step pending since compilation is blocked by: https://github.com/rapidsai/cudf/pull/7563, so tagging this one as blocked.

File: python/cudf/requirements.txt
Comment (kkraus14) [inline python/cudf/requirements.txt:1]: It looks like there's a bunch of things here that are for dev tooling, like clang, flake8, black, etc.

Can we trim this down to the minimal set of things needed to build and run cuDF from source?

File: docker_build/Dockerfile
Comment (kkraus14) [inline docker_build/Dockerfile:2]: Can we bump to CUDA 11.2 just to make sure that works as expected?

File: docker_build/Dockerfile
Comment (kkraus14) [inline docker_build/Dockerfile:9]: I'm surprised we still need these, especially with cuda in a standard location.

@gmarkall any idea here?

File: docker_build/Dockerfile
Comment (kkraus14) [inline docker_build/Dockerfile:13]: Are these actually needed? I imagine they would be the default anyway.

File: docker_build/Dockerfile
Comment (kkraus14) [inline docker_build/Dockerfile:65]: Does Arrow not have a mechanism to handle this automatically as part of the Arrow build? This is kinda gross

File: docker_build/Dockerfile
Comment (kkraus14) [inline docker_build/Dockerfile:71]: We shouldn't need the test requirements here since we're not running Arrow tests?

File: docker_build/Dockerfile
Comment (kkraus14) [inline docker_build/Dockerfile:72]: Is there a need to `--bundle-arrow-cpp` here? Looks like the arrow cpp libs are getting installed already which should allow pyarrow to use it as expected without duplicating it into the python install.

File: docker_build/Dockerfile
Comment (kkraus14) [inline docker_build/Dockerfile:76]: We should probably create a `requirements.txt` for rmm similar to what you've done for `cudf` here. We may also want to split them into `requirements.txt` (run), `requirements-build.txt` (build), and `requirements-test.txt` (test).

File: docker_build/Dockerfile
Comment (kkraus14) [inline docker_build/Dockerfile:49]: It seems like there's a lot of duplicates between pip and apt here, i.e. `cmake`, `clang`, etc.

Can we deduplicate?

File: docker_build/Dockerfile
Comment (kkraus14) [inline docker_build/Dockerfile:22]: Is this environment variable used for build or runtime?

File: docker_build/Dockerfile
Comment (kkraus14) [inline docker_build/Dockerfile:53]: What is this ppa used for?

File: docker_build/Dockerfile
Comment (gmarkall) [reply docker_build/Dockerfile:9]: Numba has been ignoring these environment variables since 0.45 - this is one of the locations it looks for these libraries anyway.

File: docker_build/Dockerfile
Comment (mike-wendt) [reply docker_build/Dockerfile:2]: FYI post CUDA 11.1 the CUDA images in `nvidia/cuda` now require the use of the patch version. So this would need to be `11.2.2` for it to work

File: docker_build/Dockerfile
Comment (mike-wendt) [inline docker_build/Dockerfile:3]: IIRC this will only need the major.minor version here `11.2`

File: docker_build/Dockerfile
Comment (mike-wendt) [reply docker_build/Dockerfile:53]: This was to get 3.7 into a 16.04 install, the changes above pull a `nvidia/cuda:11.2.2-devel-ubuntu20.04` image so this is probably not needed

File: cpp/cmake/Modules/StringifyJITHeaders.cmake
Comment (kkraus14) [inline cpp/cmake/Modules/StringifyJITHeaders.cmake:37]: This may need changes after #7583 is merged

**Review by devavret — APPROVED** [review]
Reviewing just the Jitify changes

File: docker_build/Dockerfile
Comment (galipremsagar) [reply docker_build/Dockerfile]: Looks like to pull in the docker image we would have to use patch version too `11.2.2` works but `11.2` doesn't.

File: docker_build/Dockerfile
Comment (galipremsagar) [reply docker_build/Dockerfile]: I have check all version of cuda devel & devel-ubuntu20.04 images, it seems python isn't installed out of the box, so we would need this to install python for all images.

File: docker_build/Dockerfile
Comment (galipremsagar) [reply docker_build/Dockerfile:26]: Build

File: docker_build/Dockerfile
Comment (galipremsagar) [reply docker_build/Dockerfile]: Yep removed this duplication of cpp build.

File: docker_build/Dockerfile
Comment (galipremsagar) [reply docker_build/Dockerfile]: Not needed, removed them.

File: docker_build/Dockerfile
Comment (galipremsagar) [reply docker_build/Dockerfile]: Done 👍

File: docker_build/Dockerfile
Comment (galipremsagar) [reply docker_build/Dockerfile]: Disabled building `gandiva` for arrow, we'd have to install multiple llvm versions manually if we want to build it and arrow doesn't seem to be handling it automatically.

File: python/cudf/requirements/cuda-10.1/requirements-run.txt
Comment (kkraus14) [inline python/cudf/requirements/cuda-10.1/requirements-run.txt:10]: Is this supposed to be on top of the build requirements? I think we can safely remove `notebook`, `ipython`, `partd`, `dask`, `distributed`, and `streamz` from here as those are for `dask-cudf` / `custreamz` which should land in a separate place.

File: python/cudf/requirements/requirements-build.txt
Comment (kkraus14) [inline python/cudf/requirements/requirements-build.txt:25]: I think many of these can be removed as they're only used for development for docs / linting / etc.
```suggestion
numpy
pandas>=1.0,<1.3.0dev0
cython>=0.29,<0.30
typing_extension
flatbuffers
packaging
protobuf
nvtx>=0.2.1
cachetools
numba>=0.49.0,!=0.51.0
```

File: python/cudf/pyproject.toml
Comment (kkraus14) [inline python/cudf/pyproject.toml:17]: Pretty sure a bunch of these aren't needed by the build system. Maybe just `wheel`, `setuptools`, `cython`, and maybe `protobuf`?

File: python/cudf/requirements/cuda-10.1/dev_requirements.txt
Comment (kkraus14) [inline python/cudf/requirements/cuda-10.1/dev_requirements.txt:3]: Do we need to include pyarrow in here? I know we can't pip install the cuda enabled version, but not sure if there's a mechanism to indicate it's needed.

Comment (galipremsagar) [reply python/cudf/requirements/cuda-10.1/dev_requirements.txt:3]: Should we add a comment in this file to install pyarrow too separately? Since this is something used by a developer they should be able to follow the url and install from source.

Comment (kkraus14) [reply python/cudf/requirements/cuda-10.1/dev_requirements.txt:3]: I guess? I was hoping there was a syntax for "Hey this needs to exist already, otherwise fail", but I don't see anything.

File: python/dask_cudf/requirements/cuda-10.1/dev_requirements.txt
Comment (galipremsagar) [reply python/dask_cudf/requirements/cuda-10.1/dev_requirements.txt]: Sounds good to me, changed it to that way.

File: python/cudf/requirements/cuda-10.1/dev_requirements.txt
Comment (galipremsagar) [reply python/cudf/requirements/cuda-10.1/dev_requirements.txt:6]: Added a comment here..

File: python/cudf/setup.cfg
Comment (galipremsagar) [reply python/cudf/setup.cfg]: Done

File: python/cudf/pyproject.toml
Comment (galipremsagar) [reply python/cudf/pyproject.toml]: Yup removed and verified the build end-to-end with new set of changes.

File: python/cudf/pyproject.toml
Comment (kkraus14) [inline python/cudf/pyproject.toml:10]: I'm surprised we need `cachetools` or `packaging` here as well. I figured those would be runtime only dependencies.

Comment (galipremsagar) [reply python/cudf/pyproject.toml]: We actually don't need these, removed them

**Comment by galipremsagar** [comment]
@gpucibot merge

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
@gpucibot merge

**Comment by galipremsagar** [comment]
@gpucibot merge

File: conda/environments/cudf_dev_cuda11.1.yml
Comment (ajschmidt8) [inline conda/environments/cudf_dev_cuda11.1.yml:45]:
```suggestion
  - dask>=2021.3.1
```
It looks like the `dask` version was recently updated in https://github.com/rapidsai/cudf/pull/7740. Should this be updated to match?

File: conda/environments/cudf_dev_cuda11.2.yml
Comment (ajschmidt8) [inline conda/environments/cudf_dev_cuda11.2.yml:45]:
```suggestion
  - dask>=2021.3.1
```
It looks like the `dask` version was recently updated in https://github.com/rapidsai/cudf/pull/7740. Should this be updated to match?

**Review by ajschmidt8 — APPROVED** [review]

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Pull request merged**
